include/mesh.h

```diff
@@ -32,59 +32,50 @@ class Mesh {
+                        const std::vector<double>& elevations,
+                        const std::vector<double>& demands,
+                        const std::vector<double>& leak_diameters);
+
+  //! Create Reservoir pointers
+  //! \param[in] ids reservoir id
+  //! \param[in] heads base head for reservoirs
+  bool create_reservoirs(const std::vector<Index>& ids,
+                         const std::vector<double>& heads);
+
+  //! Create Pipe pointers
+  //! \param[in] nodeids pair of end node ids for the pipe
+  //! \param[in] length, length of the pipe
+  //! \param[in] diameter, diameter of the pipe
+  //! \param[in] roughness, roughness of the pipe
+  //! \param[in] status, status of the pipe (open or close)
+  bool create_pipes(const std::vector<Index>& ids,
+                    const std::vector<std::pair<Index, Index>>& nodeids,
+                    const std::vector<double>& length,
                     const std::vector<double>& diameter,
                     const std::vector<double>& roughness,
-                    const std::vector<bool>& pipe_status);
-
-  //! Return the number of nodes in the mesh
-  //! \retval nodes_.size() number of nodes
-  unsigned nnodes() const { return nodes_.size(); }
-
-  //! Return the number of pipes in the mesh
-  //! \retval pipes_.size() number of pipes
-  unsigned npipes() const { return pipes_.size(); }
+                    const std::vector<Pipe_status>& status);
 
-  //! Remove unconnected nodes from the mesh
-  void remove_unconnected_nodes();
-
-  //! Initialize discharges in pipes
-  void initialize_pipe_discharge(
-      const std::vector<std::pair<Index, double>>& init_discharge =
-          std::vector<std::pair<Index, double>>());
-  void initialize_pipe_discharge(double init_discharge);
-
-  //! Assign initial heads/elevation for nodes
-  //! \param[in] node_head vector of pair of nodal index and initial nodal head
-  void assign_node_head(const std::vector<std::pair<Index, double>>& node_head);
-  void assign_node_elevation(
-      const std::vector<std::pair<Index, double>>& node_head);
-
-  //! Assign initial demand for nodes that have known discharge
-  //! \param[in] node_discharge vector of pair of nodal index and initial
-  //! discharge
-  void assign_node_demand(
-      const std::vector<std::pair<Index, double>>& node_discharge);
-
-  //! Make MatrixAssembler a friend class of mesh
-  friend class MatrixAssembler;
+  //! Print summary for the mesh
+  void print_summary();
 
  private:
   //! mesh id
   unsigned id_{std::numeric_limits<unsigned>::max()};
   //! nodal id and corresponding nodal pointer
   std::map<Index, std::shared_ptr<pipenetwork::Node>> nodes_;
   //! pipe id and corresponding pipe pointer
-  std::map<Index, std::shared_ptr<pipenetwork::Pipe>> pipes_;
+  std::map<Index, std::shared_ptr<pipenetwork::Link>> pipes_;
+  //! nodal id and corresponding nodal pointer
+  std::map<Index, std::shared_ptr<pipenetwork::Node>> connected_nodes_;
```

Comment:
What's the difference with a regular node?

Reply:
connected_nodes are nodes that are connected by pipes. There might be isolated nodes